File: Java-Core/src/main/java/presto/expression/ExecuteExpression.java
package presto.expression;

import presto.error.PrestoError;
import presto.error.SyntaxError;
import presto.parser.ISection;
import presto.parser.Section;
import presto.runtime.Context;
import presto.type.IType;
import presto.utils.CodeWriter;
import presto.value.CodeValue;
import presto.value.IValue;

public class ExecuteExpression extends Section implements IExpression, ISection {

	String name;
	
	public ExecuteExpression(String name) {
		this.name = name;
	}

	public String getName() {
		return name;
	}
		
	@Override
	public void toDialect(CodeWriter writer) {
		switch(writer.getDialect()) {
		case E:
			writer.append("execute: ");
			writer.append(name);
			break;
		case O:
		case S:
			writer.append("execute(");
			writer.append(name);
			writer.append(")");
			break;
		}
	}
	
	@Override
	public IType check(Context context) throws SyntaxError {
		try {
			IValue value = context.getValue(name);
			if(value instanceof CodeValue)
				return ((CodeValue) value).check(context);
			else
				throw new SyntaxError("Expected code, got:" + value.toString());
		} catch(PrestoError e) {
			throw new SyntaxError(e.getMessage());
		}
	}
	
	@Override
	public IValue interpret(Context context) throws PrestoError {
		IValue value = context.getValue(name);
		if(value instanceof CodeValue)
			return ((CodeValue) value).interpret(context);
		else
			throw new SyntaxError("Expected code, got:" + value.toString());
	}
	
}

File: Java-CodeServer/src/main/resources/js/prompto/expression/ExecuteExpression.js
var Section = require("../parser/Section").Section;
var CodeExpression = require("../expression/CodeExpression").CodeExpression;
var PrestoError = require("../error/PrestoError").PrestoError;
var SyntaxError = require("../error/SyntaxError").SyntaxError;

function ExecuteExpression(name) {
	Section.call(this);
	this.name = name;
	return this;
}


ExecuteExpression.prototype  = Object.create(Section.prototype);
ExecuteExpression.prototype.constructor = ExecuteExpression;

ExecuteExpression.prototype.toString = function() {
	return "execute: " + this.name;
};

ExecuteExpression.prototype.toDialect = function(writer) {
    writer.toDialect(this);
};

ExecuteExpression.prototype.toEDialect = function(writer) {
    writer.append("execute: ");
    writer.append(this.name);
};

ExecuteExpression.prototype.toODialect = function(writer) {
    writer.append("execute(");
    writer.append(this.name);
    writer.append(")");
};

ExecuteExpression.prototype.toSDialect = function(writer) {
    this.toODialect(writer);
};

ExecuteExpression.prototype.check = function(context) {
	try {
		var value = context.getValue(this.name);
		if(value instanceof CodeExpression) {
			return value.checkCode(context);
		} else {
			throw new SyntaxError("Expected code, got:" + value.toString());
		}
	} catch(e) {
		if(e instanceof PrestoError) {
			throw new SyntaxError(e.message);
		}
	}
};

ExecuteExpression.prototype.interpret = function(context) {
	var value = context.getValue(this.name);
	if(value instanceof CodeExpression) {
		return value.interpretCode(context);
	} else {
		throw new SyntaxError("Expected code, got:" + value.toString());
	}
};

exports.ExecuteExpression = ExecuteExpression;
